streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -337,8 +338,8 @@ public void initializeTopology() {
         if (eosEnabled) {
             try {
                 this.producer.beginTransaction();
-            } catch (final ProducerFencedException fatal) {
-                throw new TaskMigratedException(this, fatal);
+            } catch (final ProducerFencedException | UnknownProducerIdException e) {
+                throw new TaskMigratedException(this, e);
```

Comment:
I don't think that an `UnknownProducerIdException` implies that a task was migrated.

Reply:
Agreed, but `TaskMigratedException` is the mechanism to initiate a rebalance, which is currently the only way that we can re-initialize the producers.

I think we can stand to re-evaluate how we're handling all these different producer and consumer exceptions (with and without eos). Right now, there are a _lot_ of catch blocks for specific exceptions sprinkled throughout the codebase, and we universally either crash the thread or initiate a rebalance to attempt recovery. It would certainly benefit from developing a holistic approach, but right now, I'm just trying to get Streams threads to quit dying within a few hours of running in EOS mode, and I'm trying to restrain myself from broad refactoring so we can hope to merge this in for 2.4.0.

(About that last statement, I've been on holiday this week, so I haven't had time to document the comprehensive analysis in the ticket, but I do plan to make a case that it is a regression. When I ran 2.3 under similar conditions, we still had threads dying, but they were only dying from KIP-360-related-causes. In 2.4, Streams threads are dying from multiple new causes, related to the changes that we made in 2.4... Anyway, more details coming early next week, then we can discuss it)

Reply:
Give the fix for the root cause of `UnknownProducerIdException` (via KIP-360), I am wondering if we should actually treat it as fatal? I also don't see why it would be regression if KS dies if this exception is thrown though? (I understand that we did introduce some regression bugs, but those seem unrelated to `UnknownProducerIdException`?)

To be fair, 2.4.0 does not contain a full implementation of KIP-360 and thus, this fix might be a workaround... But frankly, I am not very happy about it, even if I understand the desire to just stabilize it somehow without a refactoring that won't make it into 2.4.0 anyway.

Reply:
Ok, then it seems worth hashing it out. Can you elaborate what is there to be unhappy about? I figured it would be reasonable to treat it as "recoverable by creating a new Producer" instead of "fatal and you should immediately terminate the thread".

What I was thinking was that is seems like the broker is telling us it doesn't know who we are (anymore). Because of the context, we know that it _used_ to know who we were, so it must have forgotten somehow, and it doesn't really matter how because we can always shrug it off and re-create our producer to try again.

IIUC, KIP-360 would reduce the occurrence of the exception by keeping your producer id cached even after its ttl has expired, but it by no means guarantees that it'll remember you regardless of how long you're silent.

From that perspective it seems reasonable to catch this exception and conclude, "Oops, it looks like we were silent too long and our Producer has effectively expired. We should make a new one and try again." This isn't the same thing as getting fenced, but the resolution is the same (make a new Producer and try again).

Then again, this perspective may be based on a faulty understanding of the system. Is this approach unsafe, and we should actually terminate the application instead?

Reply:
For this type or error, before KIP-360, the producer should self-recover and never throw an exception. (Cf "motivation" section of the KIP). If this error occurs, something really bad happened for the transaction and we are in an unknown state -- not sure if rebalancing and retrying is the right approach for this case \cc @guozhangwang @hachikuji @bob-barrett -- also, maybe it's just sufficient to close the task locally and recreate it -- I don't see a need to trigger a rebalance (in case it is ok to recorve from within KS).

>  This isn't the same thing as getting fenced, but the resolution is the same (make a new Producer and try again).

I disagree. If we got fenced, a new producer with the same transactional.id was created and we know that we don't own the task any longer. For a `UnknownProducerIdException` we still own the task.

> Is this approach unsafe, and we should actually terminate the application instead?

I don't thinks it's unsafe, but it's not the "right" fix IMHO, and this bug is also not a regression.

Reply:
Thanks for the clarification, @mjsax .

I agree that it would be better to just close the task and re-open it, but I don't feel comfortable introducing a whole new task lifecycle at this point for the 2.4.0 release. My rationale was that triggering a rebalance would put Kafka Streams through a well-known and well-tested code path that will result in the task getting closed and then opened again.

I do think it makes sense to go ahead and handle this case along with the two regressions. In my testing, this exception was just as fatal for Streams as the other two. Specifically, unless I include all three fixes, my soak test saw StreamThreads start dying within a few hours.

It seems like maybe a good approach right now would be to take the simple and sub-optimal path of just rebalancing when we encounter this exception for the 2.4.0 release, and file a Jira to optimize it in the way you suggest.

It would be good, by the way, to find out the answer to your question of whether it's safe to continue or not. Note that the current behavior, both in 2.4 and 2.3, without this patch is that a thread will get this exception and then shut itself down. This leads to a rebalance (since a member has left the group), after which the task is assigned to another thread, which continues processing it. Thus, we are already responding to this condition by rebalancing. It's just that we permanently lose a thread in the process. With this patch, we still rebalance, but we get to keep the thread running. If it's really unsafe to continue processing, though, we should stop the entire Streams application and force the operator to diagnose the problem with the brokers and start the app back up in a safe state.

Reply:
Thanks. That will be a good opportunity to clean up what is honestly a pretty ham-handed approach here to gracefully deal with this condition without changing too much code.